tweet_recommendations/other_methods/dbscan_based_method.py

```diff
@@ -0,0 +1,94 @@
+from collections import namedtuple
+from itertools import chain
+from typing import *
+
+import numpy as np
+import pandas as pd
+from scipy.spatial.distance import cosine
+from sklearn.cluster import DBSCAN
+from sklearn.neighbors.kd_tree import KDTree
+
+from ..estimator import Estimator
+
+CentroidData = namedtuple("CentroidData", ["cluster_medoid", "popular_hashtags"])
+
+
+class DBScanBasedEstimator(Estimator):
+    def __init__(self,
+                 verbose: bool = False):
+        self._clusters: np.ndarray = None
+
+        self.clusterizer = None
+        self.neighbours = None
+        self.centroids_data = {}
+
+        self.verbose = verbose
+
+    def fit(self, x: pd.DataFrame, y: Optional[pd.DataFrame] = None, **fit_params):
+        """
+        Builds hashtags representations using w2v method and clusters them using DBScan.
+        :param x: pd.DataFrame with tweet content, user id, and separate hashtags.
+                It is "original_tweets.p" in our case.
+        :param y: None, needed for compatibility.
+        :param fit_params:
+            minimal_hashtag_occurrence: int. If hashtag occurred less than this number
+            then it's not considered during prediction (simply removed).
+            To include all hashtags put number <= 0.
+        :return: self.
+        """
+        minimal_hashtag_occurence = fit_params["minimal_hashtag_occurence"]
+
+        x = self.drop_tweets_with_hashtags_that_occurred_less_than(x, minimal_hashtag_occurence)
```

Comment:
Czy to powinien być element tej klasy, czy raczej krok w preprocessingu?

Reply:
To zależy, jak bardzo chcemy wprowadzać granulację w klasach. Imho dla tak specjalistycznej metody nie ma co się specjalnie zmóżdżać nad kolejną klasą. Spowodowałoby to to, że dla każdej metody byśmy musieli używać `Pipeline` z sklearna, żeby połączyć ze sobą te kroki.

Reply:
Zgadzam się z @kacper1095. że jednak lepiej zostawić to w tym miejscu, tylko zrobić porządek z tą metodą, która jest wołana.